src/engine/GroupByImpl.cpp

```diff
@@ -1698,49 +1711,47 @@ IdTable GroupByImpl::createResultFromHashMap(
-       const GroupByImpl::GroupLookupResult& groupLookupResult) {
-      return CPP_template_lambda(blockSize, evaluationContext,
-                                 &groupLookupResult)(typename T, typename A)(
-          T && singleResult, A & aggregationDataVector)(
-          requires sparqlExpression::SingleExpressionResult<T> &&
-          VectorOfAggregationData<A>) {
-        // Create pairs of [value, rowIndex] from generator
-        auto generator = sparqlExpression::detail::makeGenerator(
-            std::forward<T>(singleResult), blockSize, evaluationContext);
-
-        const auto& matches = groupLookupResult.matchedRowsToGroups_;
-        // Ensure matches are sorted by rowIndex, as we will iterate
-        // through the generator in ascending order of rowIndex
-        AD_EXPENSIVE_CHECK(ql::ranges::is_sorted(
-            matches.begin(), matches.end(), {},
-            &GroupByImpl::GroupLookupResult::RowToGroup::rowIndex_));
-
-        // Process each matching row
-        // TODO<Benke> Potentially a waste of resources!
-        //   The code evaluates the sub-expressions for the non-matching rows
-        //   twice, because here they are evaluated then discarded, and later on
-        //   they are evaluated again from the materialized and sorted result.
-        auto matchIt = matches.begin();
-        size_t currentRow = 0;
-        for (const auto& value : generator) {
-          if (matchIt != matches.end() && currentRow == matchIt->rowIndex_) {
-            aggregationDataVector.at(matchIt->groupIndex_)
-                .addValue(value, evaluationContext);
-            ++matchIt;
-          }
-          ++currentRow;
-        }
-      };
-    };
+constexpr auto GroupByImpl::makeProcessGroupsVisitor(
+    size_t blockSize,
+    const sparqlExpression::EvaluationContext* evaluationContext,
+    const GroupLookupResult& groupLookupResult) {
+  return CPP_template_lambda(blockSize, evaluationContext, &groupLookupResult)(
+      typename T, typename A)(T && singleResult, A & aggregationDataVector)(
+      requires sparqlExpression::SingleExpressionResult<T> &&
+      VectorOfAggregationData<A>) {
+    const auto& matchedRows = groupLookupResult.matchedRows_;
+    const auto& groupIndexes = groupLookupResult.groupIndexes_;
+    AD_CORRECTNESS_CHECK(matchedRows.size() == groupIndexes.size());
+    // Ensure matched row indices are sorted, as we will iterate through the
+    // generator in ascending order of row index.
+    AD_EXPENSIVE_CHECK(ql::ranges::is_sorted(matchedRows));
+
+    if (matchedRows.empty()) {
+      return;
+    }
+
+    // Create a generator that only yields values for the matching row
+    // indices, skipping evaluation for non-matching rows where possible.
+    auto indices = matchedRows;
+
+    auto generator = sparqlExpression::detail::makeGeneratorSparse(
+        std::forward<T>(singleResult), blockSize, evaluationContext,
+        std::move(indices));
```

Comment:
1. no need to do a copy of the indices here. your `makeGeneratorSparse` should take some view thing (let us discuss this in person).